module/optee/voltd_regulator/src/mod_optee_voltd_regulator.c

```diff
@@ -0,0 +1,366 @@
+                                           int *level_uv)
+{
+    struct optee_voltd_regulator_dev_ctx *ctx;
+
+    ctx = find_ctx(dev_id);
+    if (ctx == NULL) {
+        return FWK_E_PARAM;
+    }
+
+    if (ctx->regulator == NULL) {
+        /* Treat unexposed voltage domain a stubbed 0V fixed level regulator */
+        *level_uv = 0;
+
+        return FWK_SUCCESS;
+    }
+
+    *level_uv = regulator_get_voltage(ctx->regulator);
+
+    FWK_LOG_DEBUG(
+        MOD_PREFIX "Get level for %u/%s: %duV",
+        fwk_id_get_element_idx(dev_id), regulator_name(ctx->regulator),
+        *level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static int optee_voltd_regulator_set_level(fwk_id_t dev_id,
+                                           int level_uv)
+{
+    struct optee_voltd_regulator_dev_ctx *ctx;
+
+    ctx = find_ctx(dev_id);
+    if (ctx == NULL) {
+        return FWK_E_PARAM;
+    }
+
+    if (ctx->regulator == NULL) {
+        return FWK_E_ACCESS;
+    }
+
+    if (regulator_set_voltage(ctx->regulator, level_uv) != TEE_SUCCESS) {
+        return FWK_E_DEVICE;
+    }
+
+    FWK_LOG_DEBUG(
+        MOD_PREFIX "Set level for %u/%s: %duV",
+        fwk_id_get_element_idx(dev_id), regulator_name(ctx->regulator),
+        level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static int optee_voltd_regulator_get_info(fwk_id_t dev_id,
+                                        struct mod_voltd_info *info)
+{
+    struct optee_voltd_regulator_dev_ctx *ctx;
+
+    ctx = find_ctx(dev_id);
+    if (ctx == NULL) {
+        return FWK_E_PARAM;
+    }
+
+    memset(info, 0, sizeof(*info));
+
+    if (!ctx->regulator) {
+        /* Treat unexposed voltage domain a stubbed 0V fixed level regulator */
+        static const char reserved[] = "reserved";
+
+        info->name = reserved;
+        info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
+        info->level_range.level_count = 1;
+        info->level_range.min_uv = 0;
+        info->level_range.max_uv = 0;
+    } else {
+        struct regulator_voltages *desc;
+
+        if (regulator_supported_voltages(ctx->regulator, &desc) != TEE_SUCCESS) {
+            return FWK_E_SUPPORT;
+        }
+
+        if (desc->type == VOLTAGE_TYPE_FULL_LIST) {
+            info->name = regulator_name(ctx->regulator);
+            info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
+            info->level_range.level_count = desc->num_levels;
+            info->level_range.min_uv = desc->entries[0];
+            info->level_range.max_uv = desc->entries[desc->num_levels - 1];
+        } else if (desc->type == VOLTAGE_TYPE_INCREMENT) {
+            info->name = regulator_name(ctx->regulator);
+            info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
+            info->level_range.min_uv = desc->entries[0];
+            info->level_range.max_uv = desc->entries[1];
+            info->level_range.step_uv = desc->entries[2];
+        } else {
+            fwk_assert(0);
```

Comment:
you can replace it for `fwk_unreachable()`

Reply:
right, thanks.